Zero memory in CRYPTO_secure_malloc.

paulidale opened the pull request:
The destroys the free list pointers that would otherwise be present in the return memory block.  This in turn help prevent information leakage from the secure memory area.

Fixes: #6989

- [x] tests are added or updated

mspncp commented on test/secmemtest.c:
Maybe you could explicitely countercheck here that the header does not consist of zero bytes only?

paulidale replied:
The header generally won't consist of zero bytes, it contains pointers to the rest of the free list.  However, if it ends up as the tail of the chain then it can have zero bytes.

I'd prefer to skip any checks than to verify that they are non-zero.  The test will be more robust for doing so.

bernd-edlinger replied:
yes, and more constant time

mspncp commented:
Shouldn't this be backported to 1.1.0? My approval holds for master and 1.1.0.

mspncp commented:
Oh, and I forgot: your commit message contains a few typos:

'The destroys' -> 'It destroys'
'return memory block' -> 'returned memory block'
'This in turn help prevent' -> 'This in turn helps prevent'

mspncp commented:
Or 'The destroys' -> 'The commit destroys'?

paulidale commented:
I'll fix the typos.

1.1.0 seems reasonable unless anyone objects.

levitte commented on crypto/mem_sec.c:
```diff
@@ -588,6 +588,9 @@ static void *sh_malloc(size_t size)
 
+    /* zero the free list header as a precaution against information leakage */
+    memset(chunk, 0, sizeof(SH_LIST));
+
```
Would it make sense to make this a `CLEAR(ret, actual_size);` up in `CRYPTO_secure_malloc` instead, just as there is one in `CRYPTO_secure_free`?

mspncp replied:
Since `CRYPTO_secure_malloc()` now zeroes memory by default, it makes sense to also change `CRYPTO_secure_zalloc()` as follows in order to avoid a double memset:

```C
void *CRYPTO_secure_zalloc(size_t num, const char *file, int line)
{
   if (!secure_mem_initialized) {
        return CRYPTO_zalloc(num, file, line);
   }
   /* CRYPTO_secure_malloc() now zeroes memory by default */
   return CRYPTO_secure_malloc(num, file, line);
}
```

mspncp replied:
Or better, call `CRYPTO_zalloc()` from `CRYPTO_secure_malloc() ` also in the `!secure_mem_initialized` case and make those two functions identical:

```
void *CRYPTO_secure_zalloc(size_t num, const char *file, int line)
{
   /* CRYPTO_secure_malloc() now zeroes memory by default */
   return CRYPTO_secure_malloc(num, file, line);
}
```

mspncp replied:
Oh, I overlooked that @paulidale's original solution erased only `sizeof(SH_LIST)` bytes, not the entire allocated buffer. Wouldn't it be better to just leave it like this instead of moving the memset on level up, @levitte?

paulidale replied:
The buffer will be cleared on `free` and is cleared after the initial `mmap(2)`.  Thus, it will always be zero unless it is allocated (once the free list header is dealt with).  So, there should be no need to clear the entire block of memory before returning as @mspncp pointed out.

I do think it makes sense to alias `CRYPTO_secure_zalloc` to `CRYPTO_secure_malloc`.  I'll double check and update in the morning (twelve or so hours).

@levitte, the free list structure contains two pointers, using CLEAR seems like overkill.  Most compilers will optimise the memset away in favour of a store or two.  As for moving it up a level, I'd be happy to do that but I think keeping it in the lower level routine makes more sense since it is dealing with the free list header struct which the higher levels have no need to know about.

paulidale commented:
I've addressed the commit message typos.  Thanks for them.

I also added a zero the free list pointers when memory blocks are coalesced together.  Otherwise, the insides of memory blocks could contain pointers to sensitive stuff.

paulidale commented:
Travis failures were relevant but are fixed now.

mspncp commented:
Of course force-pushing was necessary to change the commit message, but still the other changes could have been fixups.

paulidale commented:
Apologies for squashing things prematurely, I figured the changes were minimal (outside of the test case).

paulidale commented on crypto/mem_sec.c:
```diff
@@ -137,11 +137,17 @@ void *CRYPTO_secure_malloc(size_t num, const char *file, int line)
+#ifdef IMPLEMENTED
+    if (!secure_mem_initialized)
+        return CRYPTO_zalloc(num, file, line);
+    return CRYPTO_secure_malloc(num, file, line);
```
This seems reasonable.  What about: `/* CRYPTO_secure_malloc() now zeroes allocations when secure memory is available */`.  I'd rather not mislead anyone about it always clearing memory, it doesn't.  I contemplated using a value other than zero for the `memset(3)` to avoid accidentally initialising things for callers.

romen replied:
I don't get this: if this is applied `CRYPTO_secure_zalloc()` will not clear the whole allocated memory, right?

I don't think that's what the developer expects!

romen replied:
> The buffer will be cleared on free and is cleared after the initial mmap(2). Thus, it will always be zero unless it is allocated (once the free list header is dealt with). So, there should be no need to clear the entire block of memory before returning as @mspncp pointed out.

I see now! Uhm, maybe it's even better if this was documented in the comments to avoid confusion!

paulidale commented on crypto/mem_sec.c:
```diff
@@ -137,11 +137,17 @@ void *CRYPTO_secure_malloc(size_t num, const char *file, int line)
+        return CRYPTO_zalloc(num, file, line);
+    return CRYPTO_secure_malloc(num, file, line);
+#else
     void *ret = CRYPTO_secure_malloc(num, file, line);
```
I'm happy to do this, although I'm not fussed either way.
If nobody suggests otherwise before my morning, I'll change it.

mspncp commented:
>  I figured the changes were minimal

Small changes are even harder to detect than big ones ;-).  But no problem, and thanks for your understanding.

paulidale commented:
Merged thanks.  1.1.0 stable conflicted in the tests so a new PR will be raised.
